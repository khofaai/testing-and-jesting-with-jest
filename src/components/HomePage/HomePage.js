import React, { Component } from 'react';

class HomePage extends Component {
	constructor(props) {
		super (props);

		this.state = {
			fields : {},
			fieldErrors : {}
		}

		this.validate = this.validate.bind(this);
		this.handleSubmit = this.handleSubmit.bind(this);
		this.onInputChange = this.onInputChange.bind(this);
	}
	
	handleSubmit (e) {
		e.preventDefault();
		console.log("Why you clicked me .. ? ");

		this.setState({
			fieldErrors: this.validate(this.state.fields)
		});

		if ( !this.state.fieldErrors ) {
			console.log("Authentified.");
		} 
	}

	onInputChange (e) {
		const fields = this.state.fields;
		const newFields = {};
		newFields[e.target.name] = e.target.value;
		this.setState({
			fields: {...fields, ...newFields}
		});
	}

	validate(formData) {   
		const errors = {};
		if (!formData.name || formData.name === '' || formData.name === null) {
			console.log(formData.name);
		  errors.name = 'Please fill in the fields.';
		}
		return errors;
	}

	render() {
		return (
			<form name="form" onSubmit={this.handleSubmit}>
				{/* <img src={logo} className="App-logo" alt="logo"/> */}
				<input 
					type="email" 
					name="login" 
					placeholder="Login" 
					// required
					// value={this.state.fields.name || ''}
					onChange={ (e) => this.onInputChange(e)}
				/>
				<br/>
				<input 
					type="password" 
					name="password" 
					placeholder="Password" 
					onChange={ (e) => this.onInputChange(e)}
					// required
				/>
				<br/>
				<button type="submit">Click me</button>
				<p className="error">
				{this.state.fieldErrors.name}
				</p>
			</form>
		);
	}
}

export default HomePage;