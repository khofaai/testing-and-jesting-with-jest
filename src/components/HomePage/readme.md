# HomePage
 description 

## inputs 

 description 

## outputs 

 description 