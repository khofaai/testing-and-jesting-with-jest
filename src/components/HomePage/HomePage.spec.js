import React from 'react';
import HomePage from './HomePage';
import { shallow, mount } from 'enzyme';
import { spy } from 'sinon';

import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

 configure({ adapter: new Adapter() });


describe('<HomePage />', () => {
	let props, wrapper
  
	beforeEach(() => {
	  props = {
		handleFormSubmit: () => {
		},
	  };
	  wrapper = shallow(<HomePage {...props} />);
	});

	it("should render initial layout", () => {
    // when
    const component = shallow(<HomePage />);
    // then
    expect(component.getElements()).toMatchSnapshot();
	});

	it('should have a `<form>` element', () => {
		expect(
		  wrapper.find('form').length
		).toBe(1);
	});

	describe('<form />', () => {
		it('`<form>` element should have a onSubmit attribute', () => {
		  expect(
			wrapper.props().onSubmit
		  ).toBeDefined();
		});
	
		it('onSubmit attribute should be of type `function`', () => {
		  expect(
			typeof wrapper.props().onSubmit === 'function'
		  ).toBe(true);
		});
	
		it('`<form>` element should have a first`<input />` element', () => {
		  expect(
			wrapper.find('form').childAt(0).type()
		  ).toBe('input');
		});

		it('`<form>` element should have a second `<input />` element', () => {
			expect(
			  wrapper.find('form').childAt(2).type()
			).toBe('input');
		});

		it('`<form>` element should have an `<input type="submit" />` element', () => {
			expect(
			  wrapper.find('form').childAt(4).type()
			).toBe('button');
		});
	
		describe('<input />', () => {
			
			it('should display an error when no value is input', () => {
				const handleFormSubmit = spy();
				wrapper = mount(<HomePage handleSubmit={handleFormSubmit} />);
				wrapper.find('form').simulate('submit');
				expect(
				  wrapper.state().fieldErrors.name
				).toBe('Please fill in the fields.');
			});

				//Login input
			it('`<input>` element should be of type `text`', () => {
			  expect(
				wrapper.find('form').childAt(0).props().type
			  ).toBe('email');
			});

			it('`<input>` element should have a placeholder attribute with value `Login`', () => {
				expect(
				  wrapper.find('form').childAt(0).props().placeholder
				).toBe('Login');
			});

				//Password input
			it('`<input>` element should be of type `password`', () => {
				expect(
				  wrapper.find('form').childAt(2).props().type
				).toBe('password');
			});
		
			it('`<input>` element should have a placeholder attribute with value `Password`', () => {
				expect(
				  wrapper.find('form').childAt(2).props().placeholder
				).toBe('Password');
			});
			
		});
	});
});