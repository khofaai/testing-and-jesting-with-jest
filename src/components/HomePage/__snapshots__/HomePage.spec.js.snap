// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render initial layout 1`] = `
Array [
  <form
    name="form"
    onSubmit={[Function]}
  >
    <input
      name="login"
      onChange={[Function]}
      placeholder="Login"
      type="email"
    />
    <br />
    <input
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
    />
    <br />
    <button
      type="submit"
    >
      Click me
    </button>
    <p
      className="error"
    />
  </form>,
]
`;
